include/gunrock/algorithms/tc.hxx

```diff
@@ -0,0 +1,152 @@
+  param_t(bool _reduce_all_triangles)
+      : reduce_all_triangles(_reduce_all_triangles) {}
+};
+
+template <typename vertex_t>
+struct result_t {
+  vertex_t* vertex_triangles_count;
+  std::size_t* total_triangles_count;
+  result_t(vertex_t* _vertex_triangles_count, uint64_t* _total_triangles_count)
+      : vertex_triangles_count(_vertex_triangles_count),
+        total_triangles_count(_total_triangles_count) {}
+};
+
+template <typename graph_t, typename param_type, typename result_type>
+struct problem_t : gunrock::problem_t<graph_t> {
+  param_type param;
+  result_type result;
+
+  problem_t(graph_t& G,
+            param_type& _param,
+            result_type& _result,
+            std::shared_ptr<gcuda::multi_context_t> _context)
+      : gunrock::problem_t<graph_t>(G, _context),
+        param(_param),
+        result(_result) {}
+
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t = typename graph_t::edge_type;
+
+  void init() override {}
+
+  void reset() override {}
+};
+
+template <typename problem_t>
+struct enactor_t : gunrock::enactor_t<problem_t> {
+  enactor_t(problem_t* _problem,
+            std::shared_ptr<gcuda::multi_context_t> _context)
+      : gunrock::enactor_t<problem_t>(_problem, _context) {}
+
+  using vertex_t = typename problem_t::vertex_t;
+  using edge_t = typename problem_t::edge_t;
+  using weight_t = typename problem_t::weight_t;
+  using frontier_t = typename enactor_t<problem_t>::frontier_t;
+
+  void prepare_frontier(frontier_t* f,
+                        gcuda::multi_context_t& context) override {
+    auto P = this->get_problem();
+    auto n_vertices = P->get_graph().get_number_of_vertices();
+
+    f->sequence((vertex_t)0, n_vertices, context.get_context(0)->stream());
+  }
+
+  void loop(gcuda::multi_context_t& context) override {
+    // Data slice
+    auto E = this->get_enactor();
+    auto P = this->get_problem();
+    auto G = P->get_graph();
+
+    auto vertex_triangles_count = P->result.vertex_triangles_count;
+    auto iteration = this->iteration;
+
+    auto intersect = [G, vertex_triangles_count] __host__ __device__(
+                         vertex_t const& source,    // ... source
+                         vertex_t const& neighbor,  // neighbor
+                         edge_t const& edge,        // edge
+                         weight_t const& weight     // weight (tuple).
+                         ) -> bool {
+      if (neighbor > source) {
+        auto src_vertex_triangles_count = G.get_intersection_count(
+            source, neighbor,
+            [vertex_triangles_count](auto intersection_vertex) {
+              math::atomic::add(&(vertex_triangles_count[intersection_vertex]),
+                                vertex_t{1});
+            });
+      }
+      return false;
+    };
+
+    // Execute advance operator on the provided lambda
+    operators::advance::execute<operators::load_balance_t::block_mapped>(
+        G, E, intersect, context);
```

Comment:
Advance can actually take the entire graph as an input and choose to not output anything, which is what you're doing here. For example, for pagerank:

```cpp
    operators::advance::execute<operators::load_balance_t::block_mapped,
                                operators::advance_direction_t::forward,
                                operators::advance_io_type_t::graph,
                                operators::advance_io_type_t::none>(
        G, E, spread_op, context);
```

Notice how the input is `operators::advance_io_type_t::graph` and the output is `operators::advance_io_type_t::none`: https://github.com/gunrock/essentials/blob/dev/include/gunrock/algorithms/pr.hxx#L148-L152

Reply:
This along with disabling internal frontiers allow you to save storage on all frontiers that you might've allocated, but in reality you're just considering the entire graph as a frontier (so why not just use the graph struct itself):

https://github.com/gunrock/essentials/blob/d1f668cdb062b1c5ef442b1eef2a1cb0405e688f/include/gunrock/algorithms/pr.hxx#L209-L213